examples/threading/threading.c

```diff
@@ -1,33 +1,127 @@
+    //acquire the mutex
+    ret = pthread_mutex_lock(thread_func_args->mutex);
+    //repeat sleep
+    if(ret != 0)
+    {
+        //return with failed completion status if there's an issue with the invocation of the mutex_lock function
+        thread_func_args->thread_complete_success = false;
+        return thread_param;
+    }
+    sleep_time.tv_sec = thread_func_args->wait_before_release.tv_sec;
+    sleep_time.tv_nsec = thread_func_args->wait_before_release.tv_nsec;
+    do
+    {
+        ret=nanosleep(&sleep_time, &remaining_time);
+        if(ret == 0)
+        {
+             break;
+        }
+        sleep_time.tv_sec = remaining_time.tv_sec;
+        sleep_time.tv_nsec = remaining_time.tv_nsec;
+    }
+    while (((remaining_time.tv_sec > 0) || (remaining_time.tv_nsec > 0)));
+    //unlock the mutex
+    ret = pthread_mutex_unlock(thread_func_args->mutex);
+    if(ret != 0)
+    {
+        //return with failed completion status if there's an issue with the invocation of the mutex_unlock function
+        thread_func_args->thread_complete_success = false;
+        return thread_param;
+    }
+    //set thread completion status and return
+    thread_func_args->thread_complete_success = true;
     return thread_param;
 }
 
-
+/*
+*  Allocate memory for thread_data, setup mutex and wait arguments, pass thread_data to created thread using threadfunc() as entry point.
+*  args:
+*       parameters required for creating the thread
+*  return:
+        true if thread created successfully, false if not
+*/
 bool start_thread_obtaining_mutex(pthread_t *thread, pthread_mutex_t *mutex,int wait_to_obtain_ms, int wait_to_release_ms)
 {
-    /**
-     * TODO: allocate memory for thread_data, setup mutex and wait arguments, pass thread_data to created thread
-     * using threadfunc() as entry point.
-     *
-     * return true if successful.
-     *
-     * See implementation details in threading.h file comment block
-     */
-    return false;
+    //malloc the thread_data structure
+    struct thread_data *thread_param;
+    thread_param = malloc(sizeof(struct thread_data));
+    if(!thread_param)
+    {
+        perror("malloc");
+        return false;
+    }
+    //populate with data from args
+    thread_param->mutex = mutex;
+    thread_param->wait_before_obtain.tv_sec = (wait_to_obtain_ms/MSEC_PER_SEC);
+    thread_param->wait_before_obtain.tv_nsec = ((wait_to_obtain_ms%MSEC_PER_SEC)*NSEC_PER_MSEC);
+    thread_param->wait_before_release.tv_sec = (wait_to_release_ms/MSEC_PER_SEC);;
+    thread_param->wait_before_release.tv_nsec = ((wait_to_release_ms%MSEC_PER_SEC)*NSEC_PER_MSEC);
+    //create the thread
+    int ret = pthread_create(&thread_param->threadid,
+                             (void*)0,
+                             threadfunc,
+                             thread_param);
+    if(ret !=0)
+    {
+        //return false if thread creation failed
+        perror("Pthread Create");
+        return false;
```

Comment:
@GuruprashanthKrishnakumar what happens to `thread_param` here?

Reply:
I should have freed thread_param here? since thread did not get created and hence thread_param won't be returned (after thread completion)